text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,745 @@
+  defaultResponse: U;
+}
+```
+
+##### Changes and Breaking Changes to messages and requests
+
+Messages and requests play an important role in structuring interactions with the Toolkit.
+Because messages will be emitted in many places, they will change frequently.
+An integrator needs to take great care when handling messages and basing implementation on them.
+A list of all messages and requests is available at [TODO: link to auto-generated list of messages].
+
+For a given message code, the following changes are considered breaking:
+
+- a backwards incompatible change to type definition of message data,
+  e.g. changing the type of a property from `string` to a `boolean` or removing a non-optional property
+- removing data from a message that previously had data attached
+- removal of a message or request
+- changing the type from request to message
+
+For the avoidance of doubt, the following changes are explicitly **not** considered breaking:
+
+- a change to the message text
+- a change to the log level
+- a change to the default response of a request
+- a change to the order messages and requests are emitted in
+- the addition of new messages and requests
+- a change to the data possibly attached to the message as long as it adheres to the same type
+- adding data to a message that previously had no data
+- a backwards compatible change to the type definition of message data, e.g. adding a new property to an interface
+
+#### IoHost
+
+Message receivers need to implement the `IIoHost` interface.
+The `IoHost` is responsible for handling all logging and interactions:
+
+```ts
+interface IIoHost {
+  /**
+   * Notifies the host of a message.
+   * The caller waits until the notification completes.
+   */
+  async notify(msg: IoMessage<T>);
+  
+  /**
+   * Notifies the host of a message that requires a response.
+   * 
+   * If the host does not return a response the suggested
+   * default response from the input message will be used.
+   */
+  async requestResponse(msg: IoRequest<T,U>): U | undefined;
+}
+```
+
+For every message, the toolkit calls `notify` or `requestResponse` is called respectively.
+
+The a generic `IoEmitter` is available to provide a familiar event emitter interface.
+This implementation will do nothing for messages and requests that don't have a registered listener, i.e. the default response will be used as required.
+
+```ts
+const io = new IoEmitter();
+io.on('deploy', async (msg: IoMessage) => { /* do something */ });
+io.once('deploy:confirm-security', async (msg: IoRequest<boolean>): boolean => {
+  // reject the deployment
+  return false;
+})
+```
+
+The `CommandLineIoHost` is available, implementing all the interaction behavior for the AWS CDK CLI [TODO: this might end up in a separate package]:
+
+```ts
+const io = new CdkCliHost({
+  isInteractive: true
+})
+```
+
+##### SdkProvider
+
+  [Alternative: Make this `CredentialsProvider` or `SdkConfigProvider` instead]
```

Comment:
This is an opportunity to try to redesign the SDK abstraction, to allow for less traumatic version updates.

Reply:
Excluded for now.